### DISCUSSION TITLE
E-invoice - invalid ModalitaPagamento field

### DISCUSSION THREAD

**Discussion by ratataplan** [fattureincloud/api#25] [Q&A] [answered]
![metodo_pagamento](https://user-images.githubusercontent.com/11733873/180451921-4ad37e93-9554-40b2-aa95-6a60398abeef.png)

Ho seguito le indicazioni della guida: https://developers.fattureincloud.it/docs/guides/invoice-creation/#6%EF%B8%8F%E2%83%A3-step-six-payments-list
paymentMethod: new PaymentMethod(
        id: 386683
),
ovviamente sostituendo al numero 386683 id del mio metodo di pagamento.
La fattura viene creata regolarmente, ma la verifica formale fallisce. Infatti:
![xml](https://user-images.githubusercontent.com/11733873/180452950-4a7834e5-aafd-4dc7-bbd1-ddc59b0bc7b6.png)

nell'XML manca il metodo di pagamento.

dopo avere creato la fattura ecco la risposta:

class CreateIssuedDocumentResponse {
  Data: class IssuedDocument {
  Id: 220761420
  Entity: class Entity {
  Id: 59414300
  Code: 
  Name: MMM SRLS
  Type: 
  FirstName: 
  LastName: 
  ContactPerson: 
  VatNumber: 04376620409
  TaxCode: 
  AddressStreet: Via Cagliari, 27
  AddressPostalCode: 47838
  AddressCity: Riccione
  AddressProvince: RN
  AddressExtra: 
  Country: Italia
  Email: 
  CertifiedEmail: 
  Phone: 
  Fax: 
  Notes: 
  DefaultVat: 
  DefaultPaymentTerms: 
  DefaultPaymentTermsType: 
  DefaultPaymentMethod: 
  BankName: 
  BankIban: 
  BankSwiftCode: 
  ShippingAddress: 
  EInvoice: 
  EiCode: M5UXCR1
  HasIntentDeclaration: 
  IntentDeclarationProtocolNumber: 
  IntentDeclarationProtocolDate: 
  CreatedAt: 
  UpdatedAt: 
}

  Type: Invoice
  Number: 6
  Numeration: 
  Date: 21/07/2022 00:00:00
  Year: 2022
  Currency: class Currency {
  Id: EUR
  Symbol: €
  ExchangeRate: 1.00000
  HtmlSymbol: 
}

  Language: class Language {
  Code: it
  Name: Italiano
}

  Subject: 
  VisibleSubject: 
  RcCenter: 
  Notes: 
  Rivalsa: 0
  Cassa: 0
  AmountCassa: 0
  CassaTaxable: 100
  AmountCassaTaxable: 272
  Cassa2: 0
  AmountCassa2: 0
  Cassa2Taxable: 100
  AmountCassa2Taxable: 272
  GlobalCassaTaxable: 100
  AmountGlobalCassaTaxable: 272
  WithholdingTax: 0
  WithholdingTaxTaxable: 100
  OtherWithholdingTax: 0
  StampDuty: 0
  **PaymentMethod: class PaymentMethod {
  Id: 1505325
  Name: Bonifico bancario**
  Type: 
  IsDefault: 
  DefaultPaymentAccount: 
  Details: System.Collections.Generic.List`1[It.FattureInCloud.Sdk.Model.PaymentMethodDetails]
  BankIban: 
  BankName: 
  BankBeneficiary: 
  EiPaymentMethod: 
}

  UseSplitPayment: False
  UseGrossPrices: False
  EInvoice: True
  EiData: class IssuedDocumentEiData {
  VatKind: 
  OriginalDocumentType: 
  OdNumber: 
  OdDate: 
  Cig: 
  Cup: 
  PaymentMethod: 
  BankName: 
  BankIban: 
  BankBeneficiary: 
  InvoiceNumber: 
  InvoiceDate: 
}

  EiCassaType: 
  EiCassa2Type: 
  EiWithholdingTaxCausal: 
  EiOtherWithholdingTaxType: 
  EiOtherWithholdingTaxCausal: 
  ItemsList: System.Collections.Generic.List`1[It.FattureInCloud.Sdk.Model.IssuedDocumentItemsListItem]
  PaymentsList: System.Collections.Generic.List`1[It.FattureInCloud.Sdk.Model.IssuedDocumentPaymentsListItem]
  Template: class DocumentTemplate {
  Id: 2821
  Name: Light Smoke
  Type: 
}

  DeliveryNoteTemplate: 
  AccInvTemplate: 
  HMargins: 15
  VMargins: 16
  ShowPayments: True
  ShowPaymentMethod: False
  ShowTotals: All
  ShowPaypalButton: False
  ShowNotificationButton: False
  ShowTspayButton: False
  DeliveryNote: False
  AccompanyingInvoice: False
  DnNumber: 
  DnDate: 
  DnAiPackagesNumber: 
  DnAiWeight: 
  DnAiCausal: 
  DnAiDestination: 
  DnAiTransporter: 
  DnAiNotes: 
  IsMarked: False
  AmountNet: 272
  AmountVat: 59,84
  AmountGross: 331,84
  AmountDueDiscount: 0
  AmountRivalsa: 0
  AmountRivalsaTaxable: 272
  AmountWithholdingTax: 0
  AmountWithholdingTaxTaxable: 272
  AmountOtherWithholdingTax: 0
  AmountOtherWithholdingTaxTaxable: 272
  AmountEnasarcoTaxable: 0
  ExtraData: 
  SeenDate: 
  NextDueDate: 21/07/2022 00:00:00
  Url: https://compute.fattureincloud.it/doc/awmvy3wel8ckfu9ozll5ifc061gh1g0k.pdf
  AttachmentUrl: 
  AttachmentToken: 
  EiRaw: 
  EiStatus: NotSent
}

}

il metodo di pagamento sembra esserci, ma non viene salvato.

A qualcuno è capitata la stessa cosa? Come si risolve?

**Comment by ratataplan** [comment]
Il messaggio di errore è questo:

Exception when calling IssuedDocumentsApi.CreateIssuedDocument: Error calling CreateIssuedDocument: {"error":{"message":"Invalid request.","validation_result":{"payment_method":["The payment method may not be greater than 4 characters.","The payment method format is invalid."]}}}
Status Code: 422

**Comment by valmoz** [comment]
Buongiorno @ratataplan 
la fatturazione elettronica richiede un passaggio ulteriore, descritto alla guida che può trovare [a questo link](https://developers.fattureincloud.it/docs/guides/e-invoice-xml-customisation#-payments-list).

Il campo Modalità Pagamento dovrebbe essere al punto 2.4.2.2. del [documento ufficiale](https://developers.fattureincloud.it/docs/guides/e-invoice-xml-customisation#-the-official-xml-structure).

Dovrà quindi andare ad utilizzare i campi ei_raw come indicato in guida.

A disposizione
Mauro

**Reply by valmoz** [reply]
Mi scusi, mi sono sbagliato.
 In realtà può tranquillamente settare il campo *ei_data.payment_method* come spiegato [a questo link](https://developers.fattureincloud.it/docs/guides/invoice-creation/#3%EF%B8%8F%E2%83%A3-step-three-e-invoice)

**Comment by ratataplan** [comment]
Ho risolto parzialmente inserendo in modo statico:

invoice.EiData = new IssuedDocumentEiData(paymentMethod: "MP05");//bonifico

mentre se facevo:
invoice.EiData = new IssuedDocumentEiData(paymentMethod: method.Name);
dove method.Name scelto dalle impostazioni è bonifico mi dava errore perché più di 4 caratteri.
Dovrò fare una tabella di conversione.

**Reply by valmoz** [reply]
Buongiorno, confermo.

Poiché questi valori possono cambiare, essendo definiti nel [documento ufficiale](https://www.fatturapa.gov.it/export/documenti/fatturapa/v1.3/Rappresentazione-tabellare-fattura-ordinaria.pdf) e non da Fatture in Cloud, abbiamo deciso di non realizzare enum e di lasciare all'utilizzatore delle SDK l'onere di controllare e inserire i valori di codifica definiti nel documento. In questo modo, potrà apportare modifiche al valore quando necessario senza bisogno di attendere un nostro rilascio delle SDK.

**Comment by ratataplan** [comment]
Non c'è nessuna proprietà che mi restituisca MP05 o le impostazioni scelte dall'utente dalla classe PaymentMethod.

**Comment by valmoz** [comment, marked as answer]
To set the Modalità Pagamento, it is necessary to set the ei_data.payment_method as explained [here](https://developers.fattureincloud.it/docs/guides/invoice-creation/#3%EF%B8%8F%E2%83%A3-step-three-e-invoice)
